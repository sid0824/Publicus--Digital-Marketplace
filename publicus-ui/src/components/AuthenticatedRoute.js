import React from "react";
import { Route, Redirect } from "react-router-dom";
import {withAuthContext} from "../AuthContext";
import { message } from 'antd';

// Renders the redirect components that redirects to the supplied path
const redirect = (path) => {
  message.warning('Login With an Account To Access this Page');
  return <Redirect
    to={path} // TODO - make dynamic based on url parameters
  />
};

// Used to redirect the user away from pages that require an account. Redirects to the login page with a warning message
export const AuthenticatedRoute = ({ component: Component, authenticated, setUserHasAuthenticated, ...rest }) =>
  <Route
    {...rest}
    render={props =>
      authenticated
        ? <Component {...props} />
        : redirect(`/login`)
    }
  />;

export default withAuthContext(AuthenticatedRoute)
