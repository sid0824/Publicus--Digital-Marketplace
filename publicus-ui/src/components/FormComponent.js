import React from "react";
import { InputTypes } from "../pages/CreateForm";
import {
  DatePicker,
  Form,
  Input,
  InputNumber,
  Radio,
  Select,
  Switch,
  TimePicker
} from "antd";
const { Option } = Select;

// The types of string inputs currently available
export const ruleTypes = ["email", "number", "url"];

// Converts the list of supplied rules into a format that the form item can understand and apply
const generateFormItemRules = rules => {
  const formattedRules = [];
  rules.forEach(rule => {
    if (ruleTypes.includes(rule)) {
      formattedRules.push({ type: rule, message: `Value must be a ${rule}` });
    }
    if (rule === "required") {
      formattedRules.push({
        required: true,
        message: `This field is compulsory`
      });
    }
  });
  return formattedRules;
};

/**
 * A generic form component used in both the complete form and create form pages. It renders a form item
 * based on the configuration given with the 'question' prop for a form belonging to the 'form' prop.
 */
class FormComponent extends React.Component {
  // Return the relevant form component based on the supplied props
  getFormComponent = () => {
    const { id, question, form } = this.props;
    const { getFieldDecorator } = form;
    switch (question.type) {
      case InputTypes.TextInput:
        return (
          <Form.Item>
            {getFieldDecorator(id, {
              rules: generateFormItemRules(question.rules)
            })(<Input style={{ width: 500 }} />)}
          </Form.Item>
        );
      case InputTypes.Select:
        return (
          <Form.Item>
            {getFieldDecorator(id, {
              rules: generateFormItemRules(question.rules)
            })(
              <Select style={{ width: 500 }}>
                {question.options.map(option => (
                  <Option key={option}>{option}</Option>
                ))}
              </Select>
            )}
          </Form.Item>
        );
      case InputTypes.NumberInput:
        return (
          <Form.Item>
            {getFieldDecorator(id, {
              rules: generateFormItemRules(question.rules),
              initialValue: question.startingNumber
            })(
              <InputNumber
                style={{ width: 500 }}
                step={question.incrementor}
                min={question.min}
                max={question.max}
              />
            )}
          </Form.Item>
        );
      case InputTypes.Switch:
        return (
          <Form.Item>
            {getFieldDecorator(id, {
              rules: generateFormItemRules(question.rules)
            })(<Switch />)}
          </Form.Item>
        );
      case InputTypes.TimePicker:
        return (
          <Form.Item>
            {getFieldDecorator(id, {
              rules: generateFormItemRules(question.rules)
            })(<TimePicker />)}
          </Form.Item>
        );
      case InputTypes.DatePicker:
        return (
          <Form.Item>
            {getFieldDecorator(id, {
              rules: generateFormItemRules(question.rules)
            })(
              <DatePicker
                {...(question.showTime
                  ? { showTime: true, format: "YYYY-MM-DD HH:mm:ss" }
                  : {})}
              />
            )}
          </Form.Item>
        );
      case InputTypes.Radio:
        return (
          <Form.Item>
            {getFieldDecorator(id, {
              rules: generateFormItemRules(question.rules)
            })(
              <Radio.Group>
                {question.radioOptions.map(option => (
                  <Radio value={option}>{option}</Radio>
                ))}
              </Radio.Group>
            )}
          </Form.Item>
        );
      default:
        return null;
    }
  };

  render() {
    return <div>{this.getFormComponent()}</div>;
  }
}

export default FormComponent;
