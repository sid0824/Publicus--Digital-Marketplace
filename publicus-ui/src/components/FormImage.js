import React from "react";

const defaultImage = "https://designshack.net/wp-content/uploads/placeholder-image.png";

// Component for rendering the image linked to a form
class FormImage extends React.Component {

  // Get the form image url from props
  getState = () => {
    const { image } = this.props;
    return {
      url: image || defaultImage
    }
  };

  state = this.getState();

  // Render default form image if an error occurs
  onError = () => {
    this.setState({
      url: defaultImage
    })
  };

  render() {
    const { url } = this.state;
    const { height, width } = this.props;
    return (
      <img
        id={"formImage"}
        alt={"formImage"}
        width={height || 270}
        height={width || 180}
        src={url}
        onError={this.onError}
      />
    )
  }

}

export default FormImage;