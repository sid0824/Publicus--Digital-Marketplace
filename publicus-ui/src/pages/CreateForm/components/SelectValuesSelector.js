import React from "react";
import { Select } from "antd";
import {withCreateFormContext} from "../context/CreateFormContext";

const { Option } = Select;

/**
 * Component used in form creation to select the values for which a 'select' form item
 * will be able to choose.
 */
export class SelectValuesSelector extends React.Component {

  state = {
    currentInput: ''
  };

  /**
   * Used to update the options for the select
   * @param options - The options for the select
   */
  onChange = (options) => {
    const { changeConfigValue, id } = this.props;
    changeConfigValue(id, 'options', options);
    this.setState({
      currentTagInput: ''
    })
  };

  /**
   * Used to save the value of the current select option being typed
   * @param currentInput - The value of the current selection option being typed
   */
  onSearch = (currentInput) => {
    this.setState({
      currentInput
    })
  };

  render() {
    const { id, config } = this.props;
    const { currentInput } = this.state;

    return (
      <Select
        mode="multiple"
        size="default"
        placeholder="Add values for the select"
        defaultValue={[]}
        value={config[id].options}
        onSearch={this.onSearch}
        onChange={this.onChange}
        style={{ width: '100%' }}
        {...(!currentInput ? { dropdownRender: () => { return <div /> }} : {})}
      >
        {currentInput ? <Option key={currentInput}>{currentInput}</Option> : null}
      </Select>
    )
  }
}

export default withCreateFormContext(SelectValuesSelector);