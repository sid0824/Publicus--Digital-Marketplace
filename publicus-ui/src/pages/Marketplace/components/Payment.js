import React, {Component} from 'react';
import {CardElement, injectStripe} from 'react-stripe-elements';
import {Button, message} from "antd";

const options = {
  style: {
    base: {
      fontSize: '16px',
      color: '#424770',
      fontFamily: 'Open Sans, sans-serif',
      letterSpacing: '0.025em',
      '::placeholder': {
        color: '#aab7c4',
      },
    },
    invalid: {
      color: '#c23d4b',
    },
  }
};

export class Payment extends Component {

  state = {
    purchasing: false,
    error: null
  };

  submit = async (e) => {
    e.preventDefault();
    const { error } = this.state;
    if (!error) {
      this.setState({
        purchasing: true
      });
      try {
        const { stripe, checkout } = this.props;
        let { token } = await stripe.createToken({name: "surveyPurchase"});
        await checkout(token.id);
        message.success('Successfully Purchased Surveys');
      } catch (err) {
        console.error(err);
        message.error(err.message)
      }
      this.setState({
        purchasing: false
      });
    }
  };

  onChange = ({ error }) => {
    if (error) {
      this.setState({
        error: error.message
      })
    } else {
      this.setState({
        error: null
      })
    }
  };

  render() {
    const { purchasing, error } = this.state;
    return (
      <div>
        <br />
        <h2 style={{ marginBottom: 12 }}>Card Details</h2>
        <CardElement onChange={this.onChange} {...options} />
        <div style={{ height: 20, marginTop: 4, color: 'red' }}>
          {!!error && error}
        </div>
        <div style={{display: 'flex', flexDirection: 'row', justifyContent: 'flex-end' }}>
          <Button
            type={"primary"}
            onClick={this.submit}
            loading={purchasing}
          >
            Purchase
          </Button>
        </div>
      </div>
    );
  }
}

export default injectStripe(Payment);