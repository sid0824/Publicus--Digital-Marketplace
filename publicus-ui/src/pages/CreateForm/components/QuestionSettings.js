import React from "react";
import { withCreateFormContext } from "../context/CreateFormContext";
import {Checkbox, Input, InputNumber, Select} from "antd";
import SelectValuesSelector from "./SelectValuesSelector";
import { capitaliseFirstLetter } from "../../../util/GeneralUtil";
import { InputTypes } from "../index";
import { ruleTypes } from '../../../components/FormComponent';

const { Option } = Select;

/**
 * Component to render the settings for each form item in the create form page
 */
export class QuestionSettings extends React.Component {

  getValues = () => {
    const { config, id } = this.props;
    let selectedInputRule = '';
    const question = config[id];
    question.rules.forEach(rule => {
      if (ruleTypes.includes(rule)) {
        selectedInputRule = rule
      }
    });
    return {
      selectedInputRule, //The value of the text input type rule
      currentRadioInput: ''
    }
  };

  state = this.getValues();

  /**
   * Changes the currently selected input rule for a text input. Only allows one to be selected at a time
   * @param newInputRule - the new rule to be selected
   */
  changeSelectedInputRule = (newInputRule) => () => {
    const { config, changeConfigValue, id } = this.props;
    const { selectedInputRule } = this.state;
    const rules = config[id].rules.filter(rule => rule !== selectedInputRule);
    if (newInputRule !== selectedInputRule) {
      rules.push(newInputRule);
      this.setState({
        selectedInputRule: newInputRule
      })
    } else {
      this.setState({
        selectedInputRule: null
      })
    }
    changeConfigValue(id, 'rules', rules);
  };

  /**
   * Resets the value of the current tag be added to empty
   */
  addRadioOption = (id) => (e) => {
    const { changeConfigValue } = this.props;
    changeConfigValue(id, 'radioOptions', e);
    this.setState({
      currentRadioInput: ''
    })
  };

  /**
   * Updates the value of the current tag input
   * @param currentTagInput - new tag input
   */
  changeRadioOption = (currentRadioInput) => {
    this.setState({
      currentRadioInput
    })
  };

  /**
   * Returns the settings for a form item based on its configuration and input type
   * @returns Form Item Settings
   */
  getInputTypeSpecificSettings = () => {
    const { id , config, changeConfigValue } = this.props;
    const { selectedInputRule, currentRadioInput } =  this.state;
    const question = config[id];
    switch(question.type) {
      case InputTypes.TextInput:
        return <React.Fragment>
          <strong>Input Type</strong>
          <br />
          {ruleTypes.map((type, index) =>
            <React.Fragment key={index}>
              <Checkbox
                style={{width: 100}}
                checked={selectedInputRule === type}
                onChange={this.changeSelectedInputRule(type)}
              >
                {capitaliseFirstLetter(type)}
              </Checkbox>
              {index && index % 2 === 1 ? <br /> : null}
            </React.Fragment>
          )}
          <br />
        </React.Fragment>;
      case InputTypes.Select:
        return <React.Fragment>
          <strong>Possible Values <span style={{ color: 'red' }}>*</span></strong>
          <SelectValuesSelector
            id={id}
          />
        </React.Fragment>;
      case InputTypes.NumberInput:
        return <React.Fragment>
          <strong>Starting Number</strong>
          <br />
          <InputNumber
            style={{ width: 223 }}
            value={question.startingNumber}
            onChange={(e) => changeConfigValue(id, 'startingNumber', e)} // TODO - also set form value to update input
            min={question.min}
            max={question.max}
          />
          <br />
          <br />
          <strong>Minimum Value</strong>
          <br />
          <InputNumber
            style={{ width: 223 }}
            value={question.min}
            onChange={(e) => changeConfigValue(id, 'min', e)}
          />
          <br />
          <br />
          <strong>Maximum Value</strong>
          <br />
          <InputNumber
            style={{ width: 223 }}
            value={question.max}
            onChange={(e) => changeConfigValue(id, 'max', e)}
          />
        </React.Fragment>;
      case InputTypes.Switch:
        return <React.Fragment>
        </React.Fragment>;
      case InputTypes.DatePicker:
        return <React.Fragment>
          <Checkbox
            style={{width: 200}}
            checked={question.showTime}
            onChange={(e) => changeConfigValue(id, 'showTime', e.target.checked)}
          >
            <strong>Show Time</strong>
          </Checkbox>
        </React.Fragment>;
      case InputTypes.Radio:
        return <React.Fragment>
          <strong>Options</strong>
          <br />
          <Select
            style={{ width: 200 }}
            mode="multiple"
            size="default"
            value={question.radioOptions}
            placeholder="Add radio button options"
            onSearch={this.changeRadioOption}
            onChange={this.addRadioOption(id)}
            {...(!currentRadioInput ? { dropdownRender: () => { return <div /> }} : {})}
          >
            {currentRadioInput ? <Option key={currentRadioInput}>{currentRadioInput}</Option> : null}
          </Select>
        </React.Fragment>;
      default:
        return null;
    }
  };

  /**
   * Changes the value of the required rule for a form item
   * @param event - ChangeEvent
   */
  changeRequiredValue = (event) => {
    const { config, changeConfigValue, id } = this.props;
    const newRules =  event.target.checked
      ? [...config[id].rules, 'required']
      : config[id].rules.filter(rule => rule !== 'required');
    changeConfigValue(id, 'rules', newRules)
  };

  render() {
    const { id , config, changeConfigValue } = this.props;
    let content = this.getInputTypeSpecificSettings();
    return (
      <div>
        {
          config[id].type !== 'switch' && config[id].type !== 'radio'
            && <React.Fragment>
            <Checkbox
              checked={config[id].rules.includes('required')}
              onChange={this.changeRequiredValue}
            >
              <strong>Required Field</strong>
            </Checkbox>
            <br />
            <br />
          </React.Fragment>
        }
        <strong>Prompt</strong>
        <Input
          value={config[id].prompt}
          onChange={(e) => changeConfigValue(id, 'prompt', e.target.value)}
        />
        <br />
        <br />
        {content}
      </div>
    )
  }
}

export default withCreateFormContext(QuestionSettings);