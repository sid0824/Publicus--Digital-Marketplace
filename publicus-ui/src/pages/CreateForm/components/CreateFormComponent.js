import React from "react";
import { withCreateFormContext } from "../context/CreateFormContext";
import { Icon, Popover } from "antd";
import QuestionSettings from "./QuestionSettings";
import FormComponent from "../../../components/FormComponent";

/**
 * Used for displaying form fields whilst creating a form. Allows for deleting of fields and altering of
 * settings.
 */
export class CreateFormComponent extends React.Component {

  /**
   * Delete the given question from the form
   * @param id of the question
   */
  deleteQuestion = (id) => () => {
    const { config, changeConfig } = this.props;
    delete config[id];
    changeConfig(config);
  };

  render() {
    const { config, id, form, moveFormItemUp, moveFormItemDown } = this.props;
    const question = config[id];
    let component = <FormComponent
      id={id}
      question={question}
      form={form}
    />;

    return (
      <div key={id}>
        <strong>{id}. </strong>
        {question.prompt}
        <br />
        <div style={{ display: 'flex', flexDirection: 'row' }}>
          <div style={{ minWidth: 500 }}>
            {component}
          </div>
          <Icon
            style={{ fontSize: 20, color: 'red', marginLeft: 10, marginTop: 10 }}
            type="delete"
            onClick={this.deleteQuestion(id)}
          />
          <Icon
            style={{ fontSize: 20, color: '#4296f5', marginLeft: 10, marginTop: 10 }}
            onClick={moveFormItemUp(id)}
            type="up"
          />
          <Icon
            style={{ fontSize: 20, color: '#4296f5', marginLeft: 10, marginTop: 10 }}
            onClick={moveFormItemDown(id)}
            type="down"
          />
          <Popover placement="right" trigger="click" content={<QuestionSettings id={id} />} title={`Settings`}>
            <Icon
              style={{ fontSize: 20, color: 'grey', marginLeft: 10, marginTop: 10  }}
              type="setting"
            />
          </Popover>
        </div>
      </div>
    );
  }
}
export default withCreateFormContext(CreateFormComponent);