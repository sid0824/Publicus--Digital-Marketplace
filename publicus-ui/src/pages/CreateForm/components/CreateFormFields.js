import React from 'react';
import {CreateFormContext} from "../context/CreateFormContext";
import {Form} from "antd";
import CreateFormComponent from "./CreateFormComponent";

/**
 * Component used to wrap all form fields in the create form page. Provides form config and altering functions
 * through context to all children components.
 */
export class CreateFormFields extends React.Component {

  /**
   * Alter a specific field in the config of the form being created
   * @param id - the id of the form item being altered
   * @param field - the field on the form item being altered
   * @param value - the new value of the field being altered
   */
  changeConfigValue = (id, field, value) => {
    const { config, changeConfig } = this.props;
    config[id][field] = value;
    changeConfig(config);
  };

  /**
   * Moves a form item up in the list
   * @param stringId - id of form component to move
   */
  moveFormItemUp = (stringId) => () => {
    const { config, changeConfig } = this.props;
    const id = parseInt(stringId);
    if (id !== 1) {
      const upper = Object.assign({}, config[id - 1]);
      config[id - 1] = Object.assign({}, config[id]);
      config[id] = upper;
      changeConfig(config);
    }
  };

  /**
   * Moves a form item down in the list
   * @param stringId - id of form component to move
   */
  moveFormItemDown = (stringId) => () => {
    const { config, changeConfig } = this.props;
    const id = parseInt(stringId);
    if (id !== Object.keys(config).length) {
      const upper = Object.assign({}, config[id]);
      config[id] = Object.assign({}, config[id + 1]);
      config[id + 1] = upper;
      changeConfig(config);
    }
  };

  render() {
    const { form, changeConfig, config } = this.props;
    const context = {
      config,
      form,
      changeConfig,
      changeConfigValue: this.changeConfigValue,
      moveFormItemUp: this.moveFormItemUp,
      moveFormItemDown: this.moveFormItemDown
    };

    return (
      <CreateFormContext.Provider value={context}>
        <h1>Form Fields</h1>
        <Form onSubmit={() => {}}>
          {Object.keys(config).length
            ? Object.keys(config)
            .sort((a, b) => {
              return (a-b <= 0)
            }).map(id => <CreateFormComponent key={id} id={id} />)
            : 'Select "Create New Field" to begin adding form fields'
          }
        </Form>
      </CreateFormContext.Provider>
    )
  }
}


export default Form.create()(CreateFormFields);