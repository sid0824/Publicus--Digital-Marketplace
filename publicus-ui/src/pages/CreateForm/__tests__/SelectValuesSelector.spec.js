import toJson from 'enzyme-to-json';
import React from "react";
import {shallow, configure} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { SelectValuesSelector } from "../components/SelectValuesSelector";
import FormConfig from '../../../fixtures/FormConfig';

configure({ adapter: new Adapter() });

const mockChangeConfigValue = jest.fn();

describe('SelectValuesSelector', () => {

  beforeEach(() => {

  });

  afterEach(() => {

  });

  test('should render select values selector', () => {
    const wrapper = shallow(
      <SelectValuesSelector
        id={3}
        config={FormConfig}
      />
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  test('should add new select value option', () => {
    const wrapper = shallow(
      <SelectValuesSelector
        id={3}
        config={FormConfig}
        changeConfigValue={mockChangeConfigValue}
      />
    );
    wrapper.instance().onChange(['option1', 'option2']);
    expect(mockChangeConfigValue.mock.calls[0]).toMatchSnapshot();
    expect(wrapper.state()).toMatchSnapshot();
  });

  test('should update current select value option', () => {
    const wrapper = shallow(
      <SelectValuesSelector
        id={3}
        config={FormConfig}
        changeConfigValue={mockChangeConfigValue}
      />
    );
    wrapper.instance().onSearch('option1');
    expect(mockChangeConfigValue.mock.calls[0]).toMatchSnapshot();
    expect(wrapper.state()).toMatchSnapshot();
  });

});