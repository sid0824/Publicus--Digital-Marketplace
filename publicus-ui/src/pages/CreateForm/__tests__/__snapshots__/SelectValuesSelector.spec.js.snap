// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectValuesSelector should add new select value option 1`] = `
Array [
  3,
  "options",
  Array [
    "option1",
    "option2",
  ],
]
`;

exports[`SelectValuesSelector should add new select value option 2`] = `
Object {
  "currentInput": "",
  "currentTagInput": "",
}
`;

exports[`SelectValuesSelector should render select values selector 1`] = `
<Select
  choiceTransitionName="zoom"
  defaultValue={Array []}
  dropdownRender={[Function]}
  mode="multiple"
  onChange={[Function]}
  onSearch={[Function]}
  placeholder="Add values for the select"
  showSearch={false}
  size="default"
  style={
    Object {
      "width": "100%",
    }
  }
  transitionName="slide-up"
  value={
    Array [
      "1",
      "2",
    ]
  }
/>
`;

exports[`SelectValuesSelector should update current select value option 1`] = `
Array [
  3,
  "options",
  Array [
    "option1",
    "option2",
  ],
]
`;

exports[`SelectValuesSelector should update current select value option 2`] = `
Object {
  "currentInput": "option1",
}
`;
